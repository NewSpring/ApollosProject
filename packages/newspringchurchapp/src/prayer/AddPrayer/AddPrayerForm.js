import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Image } from 'react-native';
import {
  ModalView,
  Switch,
  TextInput,
  Button,
  styled,
  FlexedView,
} from '@apollosproject/ui-kit';
import AddPrayerHeader from './AddPrayerHeader';

const StyledButton = styled(() => ({
  borderRadius: 0,
}))(Button);

const BottomView = styled(() => ({
  justifyContent: 'flex-end',
}))(FlexedView);

const AddPrayerForm = memo(({ imgSrc, btnLabel }) => (
  <FlexedView>
    <AddPrayerHeader imgSrc={imgSrc} />
    <TextInput editable placeholder="Start typing your prayer..." />
    <BottomView>
      <Switch label="Anonymous" />
      <StyledButton title={btnLabel} />
    </BottomView>
  </FlexedView>
));

AddPrayerForm.propTypes = {
  imgSrc: Image.propTypes,
  btnLabel: PropTypes.string,
};

AddPrayerForm.defaultProps = {
  btnLabel: 'Send prayer to community',
};

AddPrayerForm.displayName = 'AddPrayerForm';
export default AddPrayerForm;
