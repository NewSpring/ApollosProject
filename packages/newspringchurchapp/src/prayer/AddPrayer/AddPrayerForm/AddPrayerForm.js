import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { View, KeyboardAvoidingView } from 'react-native';
import { SafeAreaView } from 'react-navigation';
import { Formik } from 'formik';
import {
  ModalView,
  Switch,
  TextInput,
  Button,
  styled,
  FlexedView,
  PaddedView,
  H3,
  Avatar,
} from '@apollosproject/ui-kit';

const FlexedSafeAreaView = styled({
  flex: 1,
})(SafeAreaView);

const ShrinkingView = styled(({ theme }) => ({
  flex: 1,
  paddingTop: theme.sizing.baseUnit,
}))(KeyboardAvoidingView);

const HeaderView = styled({
  alignItems: 'center',
})(View);

const TitleView = styled(({ theme }) => ({
  marginTop: theme.sizing.baseUnit * 0.5,
  marginBottom: theme.sizing.baseUnit * 2,
}))(View);

const StyledTextInput = styled({
  height: '100%',
})(TextInput);

const StyledButton = styled({
  borderRadius: 0,
})(Button);

const BottomView = styled({
  justifyContent: 'flex-end',
})(FlexedView);

const SwitchContainer = styled({
  width: '70%',
  alignSelf: 'center',
  justifyContent: 'center',
})(PaddedView);

const AddPrayerForm = memo(
  ({ onSubmit, avatarSource, title, btnLabel, ...props }) => (
    <Formik
      initialValues={{ prayer: '', anonymous: false }}
      onSubmit={onSubmit}
    >
      {({ handleChange, handleBlur, handleSubmit, values }) => (
        <ModalView {...props}>
          <FlexedSafeAreaView forceInset={{ top: 'always', bottom: 'never' }}>
            <ShrinkingView behavior={'padding'}>
              <HeaderView>
                {/* TODO: margin will be fixed pending core #667 */}
                <Avatar
                  source={values.anonymous ? null : avatarSource}
                  size={'medium'}
                />
                <TitleView>
                  <H3>{title}</H3>
                </TitleView>
              </HeaderView>
              <FlexedView>
                <PaddedView>
                  <StyledTextInput
                    editable
                    multiline
                    placeholder="Start typing your prayer..."
                    onChangeText={handleChange('prayer')}
                    onBlur={handleBlur('prayer')}
                    value={values.prayer}
                    underline={false}
                  />
                </PaddedView>
              </FlexedView>
              <BottomView>
                <SwitchContainer>
                  <Switch
                    value={values.anonymous}
                    onValueChange={handleChange('anonymous')}
                    label={'Share Anonymously'}
                  />
                </SwitchContainer>
                <StyledButton title={btnLabel} onPress={handleSubmit} />
              </BottomView>
            </ShrinkingView>
          </FlexedSafeAreaView>
        </ModalView>
      )}
    </Formik>
  )
);

AddPrayerForm.propTypes = {
  onSubmit: PropTypes.func,
  avatarSource: PropTypes.shape({
    uri: PropTypes.string,
  }),
  title: PropTypes.string,
  btnLabel: PropTypes.string,
};

AddPrayerForm.defaultProps = {
  title: 'Ask for prayer',
  btnLabel: 'Send prayer to community',
};

AddPrayerForm.displayName = 'AddPrayerForm';

export default AddPrayerForm;
