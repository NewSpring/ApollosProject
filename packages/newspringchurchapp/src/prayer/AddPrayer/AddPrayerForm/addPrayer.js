import gql from 'graphql-tag';

export default gql`
  mutation AddPrayer(
    $campusID: String!
    $categoryID: Int!
    $text: String!
    $firstName: String!
    $lastName: String
    $isAnonymous: Boolean
  ) {
    addPublicPrayerRequest(
      CampusId: $campusID
      CategoryId: $categoryID
      Text: $text
      FirstName: $firstName
      LastName: $lastName
      IsAnonymous: $isAnonymous
    ) {
      id
      campus {
        id
      }
      categoryId
      text
      firstName
      lastName
      isAnonymous
    }
  }
`;
