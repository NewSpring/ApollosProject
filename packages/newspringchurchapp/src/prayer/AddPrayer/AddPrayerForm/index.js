import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Image, View } from 'react-native';
import { SafeAreaView } from 'react-navigation';
import { Formik } from 'formik';
import {
  ModalView,
  Switch,
  TextInput,
  Button,
  styled,
  FlexedView,
  PaddedView,
  H3,
  H5,
  Avatar,
} from '@apollosproject/ui-kit';

const FlexedSafeAreaView = styled({
  flex: 1,
})(SafeAreaView);

const HeaderView = styled({
  alignItems: 'center',
})(View);

const TitleView = styled(({ theme }) => ({
  marginTop: theme.sizing.baseUnit * 0.5,
  marginBottom: theme.sizing.baseUnit * 2,
}))(View);

const StyledTextInput = styled({
  height: '100%',
})(TextInput);

const StyledButton = styled({
  borderRadius: 0,
})(Button);

const BottomView = styled({
  justifyContent: 'flex-end',
})(FlexedView);

const SwitchContainer = styled({
  width: '70%',
  alignSelf: 'center',
  flexDirection: 'row',
  justifyContent: 'center',
})(PaddedView);

const SwitchLabel = styled(({ theme }) => ({
  color: theme.colors.primary,
}))(H5);

const SwitchLabelContainer = styled(({ theme }) => ({
  justifyContent: 'center',
  marginLeft: theme.sizing.baseUnit,
}))(View);

const AddPrayerForm = memo(({ imgSrc, title, btnLabel, ...props }) => (
  <Formik
    initialValues={{ prayer: '', anonymous: false }}
    onSubmit={(values) => console.log(values)}
  >
    {({ handleChange, handleBlur, handleSubmit, values }) => (
      <ModalView {...props}>
        <FlexedSafeAreaView forceInset={{ top: 'always', bottom: 'never' }}>
          <HeaderView>
            <Avatar source={values.anonymous ? null : imgSrc} size={'medium'} />
            <TitleView>
              <H3>{title}</H3>
            </TitleView>
          </HeaderView>
          {/* TODO: add underline={false} prop to TextInput, 
        pending https://github.com/ApollosProject/apollos-prototype/issues/629 */}
          <FlexedView>
            <PaddedView>
              <StyledTextInput
                editable
                multiline
                placeholder="Start typing your prayer..."
                onChangeText={handleChange('prayer')}
                onBlur={handleBlur('prayer')}
                value={values.prayer}
              />
            </PaddedView>
          </FlexedView>
          <BottomView>
            <SwitchContainer>
              <Switch
                value={values.anonymous}
                onValueChange={handleChange('anonymous')}
              />
              <SwitchLabelContainer>
                <SwitchLabel>Share Anonymously</SwitchLabel>
              </SwitchLabelContainer>
            </SwitchContainer>
            <StyledButton title={btnLabel} onPress={() => console.log(props)} />
          </BottomView>
        </FlexedSafeAreaView>
      </ModalView>
    )}
  </Formik>
));

AddPrayerForm.propTypes = {
  imgSrc: Image.propTypes,
  title: PropTypes.string,
  btnLabel: PropTypes.string,
};

AddPrayerForm.defaultProps = {
  title: 'Ask for prayer',
  btnLabel: 'Send prayer to community',
};

AddPrayerForm.displayName = 'AddPrayerForm';
export default AddPrayerForm;
