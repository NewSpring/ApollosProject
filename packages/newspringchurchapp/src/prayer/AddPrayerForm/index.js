import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { View, KeyboardAvoidingView } from 'react-native';
import { SafeAreaView } from 'react-navigation';
import { Formik } from 'formik';
import { get } from 'lodash';

import {
  ModalView,
  Switch,
  TextInput,
  Button,
  styled,
  FlexedView,
  PaddedView,
} from '@apollosproject/ui-kit';

import { useQuery, useMutation } from 'react-apollo';

import GET_USER_PROFILE from 'newspringchurchapp/src/tabs/connect/getUserProfile';
import GET_USER_PRAYERS from 'newspringchurchapp/src/prayer/data/queries/getUserPrayers';
import ADD_PRAYER from 'newspringchurchapp/src/prayer/data/mutations/addPrayer';

import PrayerHeader from '../PrayerHeader';

const FlexedSafeAreaView = styled({
  flex: 1,
})(SafeAreaView);

const ShrinkingView = styled(({ theme }) => ({
  flex: 1,
  paddingTop: theme.sizing.baseUnit,
}))(KeyboardAvoidingView);

const StyledTextInput = styled({
  height: '100%',
})(TextInput);

const StyledButton = styled({
  borderRadius: 0,
})(Button);

const BottomView = styled({
  justifyContent: 'flex-end',
})(FlexedView);

const SwitchContainer = styled({
  width: '70%',
  alignSelf: 'center',
  justifyContent: 'center',
})(PaddedView);

const StyledPrayerHeaderView = styled(({ theme }) => ({
  marginBottom: theme.sizing.baseUnit,
}))(View);

const AddPrayerForm = memo(({ title, btnLabel, navigation, ...props }) => {
  const userData = useQuery(GET_USER_PROFILE);

  const [addPrayer] = useMutation(ADD_PRAYER, {
    update: (cache, { data: { addPrayer: addedPrayer } }) => {
      const { userPrayers } = cache.readQuery({
        query: GET_USER_PRAYERS,
      });
      cache.writeQuery({
        query: GET_USER_PRAYERS,
        data: {
          userPrayers: userPrayers.concat([addedPrayer]),
        },
      });
    },
  });

  const avatarSource = get(userData, 'currentUser.profile.photo', {
    uri: null,
  });

  return (
    <Formik
      initialValues={{ prayer: '', anonymous: false }}
      onSubmit={({ prayer, anonymous }) => {
        addPrayer({
          variables: {
            campusId: get(userData, 'data.currentUser.profile.campus.id'),
            // TODO: make this dynamic
            categoryId: 2,
            text: prayer,
            firstName: get(userData, 'data.currentUser.profile.firstName'),
            lastName: get(userData, 'data.currentUser.profile.lastName'),
            isAnonymous: anonymous,
          },
        });

        navigation.pop();
      }}
    >
      {({ handleChange, handleBlur, handleSubmit, values }) => (
        <ModalView {...props}>
          <FlexedSafeAreaView forceInset={{ top: 'always', bottom: 'never' }}>
            <ShrinkingView behavior={'padding'}>
              <StyledPrayerHeaderView>
                <PrayerHeader
                  avatarSource={values.anonymous ? null : avatarSource}
                  avatarSize={'medium'}
                  name={title}
                />
              </StyledPrayerHeaderView>
              <FlexedView>
                <PaddedView>
                  <StyledTextInput
                    editable
                    multiline
                    placeholder="Start typing your prayer..."
                    onChangeText={handleChange('prayer')}
                    onBlur={handleBlur('prayer')}
                    value={values.prayer}
                    underline={false}
                  />
                </PaddedView>
              </FlexedView>
              <BottomView>
                <SwitchContainer>
                  <Switch
                    value={values.anonymous}
                    onValueChange={handleChange('anonymous')}
                    label={'Share Anonymously'}
                  />
                </SwitchContainer>
                <StyledButton title={btnLabel} onPress={handleSubmit} />
              </BottomView>
            </ShrinkingView>
          </FlexedSafeAreaView>
        </ModalView>
      )}
    </Formik>
  );
});

AddPrayerForm.propTypes = {
  title: PropTypes.string,
  btnLabel: PropTypes.string,
};

AddPrayerForm.defaultProps = {
  title: 'Ask for prayer',
  btnLabel: 'Send prayer to community',
};

AddPrayerForm.displayName = 'AddPrayerForm';

export default AddPrayerForm;
