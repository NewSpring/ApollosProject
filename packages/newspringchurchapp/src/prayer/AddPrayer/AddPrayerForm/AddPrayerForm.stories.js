import React from 'react';
import { storiesOf } from '@storybook/react-native';

import AddPrayerForm from './AddPrayerForm';

storiesOf('prayer/AddPrayer/AddPrayerForm', module)
  .add('default', () => <AddPrayerForm navigation={{ pop: () => {} }} />)
  .add('avatarSource', () => (
    <AddPrayerForm
      navigation={{ pop: () => {} }}
      avatarSource={{ uri: 'https://picsum.photos/55/55?random' }}
    />
  ))
  .add('btnLabel', () => (
    <AddPrayerForm
      navigation={{ pop: () => {} }}
      btnLabel={'Custom button label'}
    />
  ));
