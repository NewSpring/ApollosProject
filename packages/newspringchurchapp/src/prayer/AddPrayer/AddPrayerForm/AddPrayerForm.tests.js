import React from 'react';
import renderer from 'react-test-renderer';

import AddPrayerForm from '.';

describe('The AddPrayerForm component', () => {
  it('should render', () => {
    /* we have to pass in a date via defaultDate or the DateInput component will create a current date
     * object and invalidate the snapshots every time. */
    const tree = renderer.create(<AddPrayerForm navigation={jest.fn()} />);
    expect(tree).toMatchSnapshot();
  });
  it('should render a custom avatar', () => {
    /* we have to pass in a date via defaultDate or the DateInput component will create a current date
     * object and invalidate the snapshots every time. */
    const tree = renderer.create(
      <AddPrayerForm
        navigation={jest.fn()}
        imgSrc={{ uri: 'https://picsum.photos/55/55?random' }}
      />
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render a custom title', () => {
    /* we have to pass in a date via defaultDate or the DateInput component will create a current date
     * object and invalidate the snapshots every time. */
    const tree = renderer.create(
      <AddPrayerForm navigation={jest.fn()} btnLabel={'custom button label'} />
    );
    expect(tree).toMatchSnapshot();
  });
});
