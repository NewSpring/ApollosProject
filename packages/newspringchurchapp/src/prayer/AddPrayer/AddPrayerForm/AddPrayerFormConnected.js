import React from 'react';
import { Query, Mutation } from 'react-apollo';
import { get } from 'lodash';
import getUserProfile from 'newspringchurchapp/src/tabs/connect/getUserProfile';
import getPrayers from 'newspringchurchapp/src/prayer/data/queries/getUserPrayers';
import addPrayer from '../../data/mutations/addPrayer';
import AddPrayerForm from './AddPrayerForm';

class AddPrayerFormConnected extends React.Component {
  static navigationOptions = {
    header: null,
  };

  render() {
    return (
      <Query query={getUserProfile} fetchPolicy={'cache-only'}>
        {({ data: userData }) => (
          <Mutation
            mutation={addPrayer}
            update={(cache, { data: { addPublicPrayerRequest } }) => {
              const { getCurrentPersonPrayerRequests } = cache.readQuery({
                query: getPrayers,
              });
              cache.writeQuery({
                query: getPrayers,
                data: {
                  getCurrentPersonPrayerRequests: getCurrentPersonPrayerRequests.concat(
                    [addPublicPrayerRequest]
                  ),
                },
              });
            }}
          >
            {(createPrayer) => (
              <AddPrayerForm
                onSubmit={(values) => {
                  createPrayer({
                    variables: {
                      campusID: get(userData, 'currentUser.profile.campus.id'),
                      // TODO: make this dynamic
                      categoryID: 2,
                      text: values.prayer,
                      firstName: get(userData, 'currentUser.profile.firstName'),
                      lastName: get(userData, 'currentUser.profile.lastName'),
                      isAnonymous: values.anonymous,
                    },
                  });
                  // TODO: wonder if this should be tied to the ModalView onClose method???
                  this.props.navigation.pop();
                }}
                avatarSource={get(userData, 'currentUser.profile.photo', {
                  uri: null,
                })}
                {...this.props}
              />
            )}
          </Mutation>
        )}
      </Query>
    );
  }
}

export default AddPrayerFormConnected;
