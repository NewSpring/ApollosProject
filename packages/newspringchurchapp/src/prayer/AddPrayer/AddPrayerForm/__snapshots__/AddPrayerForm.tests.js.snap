// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AddPrayerForm component should render 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`The AddPrayerForm component should render a custom avatar 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`The AddPrayerForm component should render a custom title 1`] = `
<RNCSafeAreaView
  onInsetsChange={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
